Issue/4727 notifs detail view horizontal navigation

mzorz opened the pull request:
Fixes #4727 

This PR implements 2 ways to horizontally navigate through Notifications detail:
- using swipe left/right (by means of a ViewPager control)
- touching the left/right arrows now available on the action bar

To test:
1. log in to the WP Android app with your account
2. go to the Notifications tab (rightmost tab)
3. you'll be presented with the notifications list.
4. tap on any item -  you'll be presented with the item's detail screen
5. swipe left-right to navigate. 
6. Note that you'll be able to navigate left-right items as they are listed top-bottom in the notifications list correspondingly. That means if you tapped on the first item in the list, you won't be able to swipe left on that item (as you can't go any further up while already looking at the top of the list). Same thing for the last item in the list.
7. Also, if you tap on any of the filters (All, Comments, unread, etc.)  you'd be able to navigate the list horizontally accordingly - that is, items are filtered there as well.  

See video here https://cloudup.com/cBZOd5sFriD

cc @daniloercoli @jleandroperez @kwonye

daniloercoli commented:
I  found an issue where the details screen can be blank upon screen rotation.
Steps to repro:
- Go to notifications
- Tap on a note and open the details screen
- Rotate the device

daniloercoli commented:
Hey @nbradbury - Since you've worked with the `ViewPager` in the past :) , would you mind a quick look to this PR and tell us if you see something to double check?

nbradbury commented:
Howdy @daniloercoli, sure I'll take a look later today. In the meantime, now that I see this I think we should reconsider the left/right icons in the toolbar. 

This isn't the only area of the app that uses a ViewPager (the reader being the obvious other place), and I think we'll want a consistent UI for letting the user know they can move between pages. I don't see the left/right icons working well in the reader since there are already a couple of toolbar icons there.

I think what we really want is to simply let the user know they can swipe between pages - once they know they can do it, it's much easier than tapping an icon. I think the USA Today app handles this nicely. The first few times you open an article, a snackbar-ish message animates in at the bottom of the screen letting you know you can swipe. Once you swipe, it animates back out.

What do we think about that approach instead?

![screenshot_20161117-131144](https://cloud.githubusercontent.com/assets/3903757/20402013/024a60c2-acc9-11e6-8a1b-052f56bc6433.png)

daniloercoli commented:
I agree with you @nbradbury, and the proposed solution looks much nicer.
IMO Left/Right buttons in the toolbar seems off-platform.

Curious to gear what @drw158 thinks about it.

/cc @kwonye

nbradbury commented:
> Hey @nbradbury - Since you've worked with the ViewPager in the past :) , would you mind a quick look to this PR and tell us if you see something to double check?

It looks like this ViewPager adapter is based on the one I coded for the reader a while ago, and that has been pretty reliable. I do recommend adding [this change](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostPagerActivity.java#L780-L791), though, to address [this Android bug](https://code.google.com/p/android/issues/detail?id=42601).

Note, also, that the use of `getChildFragmentManager()` [here](https://github.com/wordpress-mobile/WordPress-Android/blob/9289ed1ffd94494662610f58bbe5eae842c75a6b/WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java#L1049) will need to be re-thought. That call was added in API 17 and our `minSdkVersion` is 16.

mzorz commented:
> Note, also, that the use of getChildFragmentManager() here will need to be re-thought. That call was added in API 17 and our minSdkVersion is 16.

That will be a tough one. I needed to use `getChildFragmentManager` because otherwise this call here https://github.com/wordpress-mobile/WordPress-Android/blob/9289ed1ffd94494662610f58bbe5eae842c75a6b/WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java#L1062  would replace fragments on the first available fragment in the ViewPager adapter, instead of doing the replacement in the correct page (for instance, in the _next_ page as the viewpager creates and pre-loads offscreen), which led to very undesirable (and unpredictable) results such as blank pages or a "next" item actually being displayed (because the content of such item would be replaced on the first View found in the previous object). I actually spent more time trying to figure out why that was happening (and finding `getChildFragmentManager` addresses exactly that kind of situation) than implementing the feature itself.

That said, I believe a workaround could be using tags for fragments (so we can search for them individually as opposed to what happens with xml ids), but we'd have to take care of fragment states and pages artificially, outside the fragment manager to circumvent this problem. 

Any thoughts are appreciated

mzorz commented:
let's blame the jet lag - looking a bit closer this should be in in the support libraries. Let me get back to this later.

mzorz commented:
re: not using the left/right arrows on the action bar, and instead using the toast/snackbar to inform the user they have the option to swipe to navigate through notifications, I love that. 

I'd definitely go that way unless @kwonye or @drw158 feel differently

mzorz commented:
Also, if we go the way @nbradbury suggests, maybe we'd like to just keep the EDIT button where it was? (check this #4782) @kwonye @drw158

daniloercoli commented:
My memory is fuzzy, but I remember we were using ChildFragments in Stats, and switched to using "normal" Android Fragments and tags.

mzorz commented:
Just opened a new PR for discussing the support library usage and related topics here #4806 .

drw158 commented:
Using a notice to inform the user that they can swipe seems like a great idea. No need for the navigation buttons in that case.

mzorz commented:
Thank you @drw158 ! Going that way then. :+1:

mzorz commented:
Okay! implemented changes as agreed:

+ used `setId` to identify each fragment container instead of relying on `ChildFragmentManager`
(also closed #4806 as per this commit 98c2884)
+ ditched the left/right arrows and get the EDIT button back on the action bar again
+ added a toast to tell the user they can swipe to navigate (with an action so it doesn't show up every time)
+ see transparent background (I guess should be ok now, but couldn’t replicate actually - set the layout background for `notifications_fragment_detail_list.xml` to light grey as the comments were that way)
+ added analytics for swipe (page changes and tap on the snackbar)
+ handling device rotation properly now

Video here: https://cloudup.com/c699B6sDvp3

@nbradbury @daniloercoli @drw158 @kwonye ready for another pass 👍 🙇

nbradbury commented:
> added a ~~toast~~ **snackbar** to tell the user they can swipe to navigate (with an action so it doesn't show up every time)

I like the snackbar, but I think we can get rid of the "OKAY" action. Rather than force the user to dismiss the snackbar to stop it from showing again, how about we just stop showing it once the user has swiped?

Also, if it's going to have an arrow to indicate the user can swipe left, it should have one to indicate when they can swipe right (or we can keep it simple and drop the arrows).

mzorz commented:
Thank you @nbradbury ! made these changes here https://github.com/wordpress-mobile/WordPress-Android/pull/4796/commits/61b4f76f1f48f5633c31c6aac7c8d2fd6de1321c

nbradbury commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java:
```diff
@@ -148,16 +157,106 @@ protected void onSaveInstanceState(Bundle outState) {
+        super.onStart();
+        //if the user hasn't used swipe yet, hint the user they can navigate through notifications detail
+        //using swipe on the ViewPager
+        if (!AppPrefs.isSwipeToNavigateShown()) {
```
Are there situations where only one note is shown, so there's no swiping? If so, we should make sure not to show this snackbar in those situations.

mzorz replied:
oops. You're so right! here's the fix ccdb2a9

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderViewPagerTransformer.java:
```diff
@@ -11,16 +11,16 @@
  */
-class ReaderViewPagerTransformer implements ViewPager.PageTransformer {
-    enum TransformType {
+public class ReaderViewPagerTransformer implements ViewPager.PageTransformer {
```
We should move this class to  `Widgets` (or `Utils`) and call it `WPViewPagerTransformer`.

mzorz replied:
Addressed in 382808f

daniloercoli commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -855,6 +855,8 @@
     <string name="notifications_label_new_notifications">New notifications</string>
     <string name="notifications_label_new_notifications_subtitle">Tap to show them</string>
+    <string name="notifications_label_swipe_for_more_snackbar">Swipe for more notifications</string>
+    <string name="notifications_label_swipe_snackbar_dont_show_again">OKAY</string>
```
Is it used?

kwonye commented:
I'll defer to everyones' judgement here: design's 👌 with me

Thanks @drw158 for the feedback!

kwonye commented on WordPress/src/main/java/org/wordpress/android/ui/notifications/NotificationsDetailActivity.java:
```diff
@@ -1,12 +1,18 @@
 import android.support.v7.app.AppCompatActivity;
 import android.text.TextUtils;
 import android.view.MenuItem;
+import android.view.View;
```
No shame in nitpicks! 😁

mzorz replied:
done :)

mzorz commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -1109,11 +1098,40 @@ public void onNoteChanged(Note note) {
 
+    private void setIdForCommentContainer(){
+        if (mComment != null && mCommentContentLayout != null) {
+            mCommentContentLayout.setId((int)mComment.commentID);
```
A note on using tags vs IDs, the fragment transaction needs the container `id` where to add the fragment to, and there doesn't seem to be a way to replace the container contents without knowing its `id` in FragmentManager (though this artifact we're making by having a different ID for each _viewpager page's container_ is handled seamlessly by the manager returned by getChildFragmentManager, which is the main purpose of its existence, i.e. handling nested fragments). Setting a tag to find it by tag later then having to set the id anyways seems dumb, so what I did is I'm generating an `id` with a base id plus the Note position index in the notes adapter array to make probability of id collision low (while it is theoretically possible, I don't think it'll ever happen in practice, as we'd have to have a list of more than 65535 items in the notifications list...)

BTW there is also an API to generate a unique id on runtime, but it's also only available since API level 17, which is what we were trying to avoid in the first place 😁 

Commit addressing all this here 8da2cea

mzorz commented:
This is ready for a  (I guess final) review @daniloercoli !
🙇

daniloercoli commented on WordPress/src/main/java/org/wordpress/android/ui/comments/CommentDetailFragment.java:
```diff
@@ -420,6 +433,13 @@ private void setNoteWithNoteId(String noteId) {
         setRemoteBlogId(note.getSiteId());
     }
 
+    private void setIdForFragmentContainer(int id){
```
What do you think if we remove this method and move the assignment `mIdForFragmentContainer = id;` in `newInstance` ? Since it's called only there, and the name of this method is quite similar to ` setIdForCommentContainer` confusing me a bit at the beginning of the review.

daniloercoli replied:
Ok, right.

daniloercoli commented on libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTracker.java:
```diff
@@ -124,6 +124,7 @@
         NOTIFICATION_UNLIKED,
         NOTIFICATION_TRASHED,
         NOTIFICATION_FLAGGED_AS_SPAM,
+        NOTIFICATION_SWIPE_PAGE_CHANGED,
```
We should bump Analytics for this new event in both `AnalyticsTrackerMixpanel` and `AnalyticsTrackerNosara`. Please   read the documentation about adding new events (or ping me in Slack) before setting a name for this event.

mzorz replied:
Thanks a lot for your explanation and directions on chat! addressed in 9651cdf  - would you take another look please? 🙇

daniloercoli commented:
I see this following crash upon device rotation:
```
11-23 13:28:27.985 25674-25674/org.wordpress.android E/AndroidRuntime: FATAL EXCEPTION: main
                                                                       Process: org.wordpress.android, PID: 25674
                                                                       java.lang.RuntimeException: Unable to resume activity {org.wordpress.android/org.wordpress.android.ui.notifications.NotificationsDetailActivity}: java.lang.IllegalArgumentException: Must use non-zero containerViewId
                                                                           at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3400)
                                                                           at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3440)
                                                                           at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2713)
                                                                           at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4483)
                                                                           at android.app.ActivityThread.-wrap19(ActivityThread.java)
                                                                           at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1466)
                                                                           at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                           at android.os.Looper.loop(Looper.java:154)
                                                                           at android.app.ActivityThread.main(ActivityThread.java:6077)
                                                                           at java.lang.reflect.Method.invoke(Native Method)
                                                                           at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                                                                           at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
                                                                        Caused by: java.lang.IllegalArgumentException: Must use non-zero containerViewId
                                                                           at android.app.BackStackRecord.replace(BackStackRecord.java:501)
                                                                           at android.app.BackStackRecord.replace(BackStackRecord.java:496)
                                                                           at org.wordpress.android.ui.comments.CommentDetailFragment.showCommentForNote(CommentDetailFragment.java:1132)
                                                                           at org.wordpress.android.ui.comments.CommentDetailFragment.showComment(CommentDetailFragment.java:612)
                                                                           at org.wordpress.android.ui.comments.CommentDetailFragment.setNote(CommentDetailFragment.java:417)
                                                                           at org.wordpress.android.ui.comments.CommentDetailFragment.setNoteWithNoteId(CommentDetailFragment.java:432)
                                                                           at org.wordpress.android.ui.comments.CommentDetailFragment.onResume(CommentDetailFragment.java:347)
                                                                           at android.app.Fragment.performResume(Fragment.java:2398)
                                                                           at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1032)
                                                                           at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1171)
                                                                           at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1153)
                                                                           at android.app.FragmentManagerImpl.dispatchResume(FragmentManager.java:2049)
                                                                           at android.app.FragmentController.dispatchResume(FragmentController.java:198)
                                                                           at android.app.Activity.performResume(Activity.java:6789)
                                                                           at android.app.ActivityThread.performResumeActivity(ActivityThread.java:3377)
                                                                           at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:3440) 
                                                                           at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2713) 
                                                                           at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:4483) 
                                                                           at android.app.ActivityThread.-wrap19(ActivityThread.java) 
                                                                           at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1466) 
                                                                           at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                           at android.os.Looper.loop(Looper.java:154) 
                                                                           at android.app.ActivityThread.main(ActivityThread.java:6077) 
                                                                           at java.lang.reflect.Method.invoke(Native Method) 
                                                                           at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865) 
                                                                           at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755) 
```

mzorz commented:
gotcha - let me check it out right away.

mzorz commented:
addressed crash in commit 6210d32

daniloercoli commented:
:shipit:
